### ISSUE TITLE
JavaGUI selectNode issues

### ISSUE THREAD

**Issue by dergo** [qspin/qtaste#163]
1/ If you try to use "." as separator, the call succeeds but doesn't select anything and the following exception occurs in the SUT:
java.lang.ArrayIndexOutOfBoundsException: 0
	at com.qspin.qtaste.javagui.server.TreeNodeSelector.prepareDoActions(TreeNodeSelector.java:95)
This is because the separator is used as a regular expression, which is not documented, so you must then use "\.", and I'm not sure it's useful to have a regexp as separator.
Also if for any reason there is an exception the call should fail.

2/ If you try to select the root node, the call succeeds but doesn't select anything and the following exception occurs in the SUT:
java.lang.IllegalArgumentException: path in TreePath must be non null and not empty.
	at javax.swing.tree.TreePath.<init>(TreePath.java:102)
	at com.qspin.qtaste.javagui.server.TreeNodeSelector.doActionsInSwingThread(TreeNodeSelector.java:120)
The selection should work and if for any reason there is an exception the call should fail.

3/ documentation is not clear:
nodeName - the node to select.
=> confusing name because actually the method uses the node label, not the node component name
=> it is not secified that we must use a path to the node, not just the node itself
nodeSeparator - the node separator used in the value parameter.
=> there is no value parameter, it should be nodeName

4/ if the label contains HTML formatting, you cannot simply pass the text to selectNode, you currently have to pass the full HTML text, e.g. "<html><i>the label</i></html>" which is not great

5/ debug System.out.println are still present in the code and should be removed or replaced by log4j debug logs

6/ typo in exceptions: "Unabled to find" => should be "Unable"

**Comment by dergo** [comment]
7/ If you try to select an inexistent node, the TreeNodeSelector will continue to loop over all tree nodes during 10 seconds, then the call will return successfully. It should loop over all tree nodes only once and the call should fail.

**Comment by remybaranx** [comment]
Here is the correction of this bug, except point (4) : https://github.com/qspin/qtaste/pull/169. 

Some details : 

(1) now, the node path separator is not considered a a regex (it is escaped using `Pattern.quote()`). So we can use `.`, `\`, etc ...

(2) If the root node is visible and you select it, it succeeds. If the root is NOT visible, and you select it, it raises a QTasteTestFailException. I think it's the expected behaviour, isn't it ?

(3) changed the parameters names and their documentation. Now, we specify a nodePath and a nodePathSeparator. Is it ok ?

(4) I understand that it's not great to specify the full HTML text but, I think, it's out of the scope of this `selectNode()` method. 
Imagine you have a red bold "toto" node and a blue italic "toto" node at the same level. We cannot select the right node only with the final text. We have to find another way (full html, node position in the tree, etc ...). It the same problem in a classic tree, if you have two nodes with the same text at the same level. How to select the right one ?
So, I think the current `selectNode()` method is not enough to cover all the cases, and we should add others `selectNode()` methods to handle others cases (selection with node positions in the tree for example).

(5) I replaced by log4j calls.

(6) ok

(7) now it works.

**Comment by dergo** [comment]
For (4), I cannot simply use full HTML text because I don't know if the label is currently formatted or not, it depends on the current state.
So I would have to try to select unformatted node, catch exception if any then select formatted node, which is not great.

Using a regular expression for the label would be a solution as I could use ".*label.*", if removing the HTML formatting is too complex.
In this case, it's probably better to add a separate method name e.g. `selectNodeRe()`.

Providing a method to cover cases where several nodes have the same name is currently not needed.

**Comment by remybaranx** [comment]
Yes, I understand your problem, and yes, I think adding new `selectNode` methods such as `selectNodeRe()` is the best way to solve this issue.

So, I'm going to add this new method.

**Comment by remybaranx** [comment]
It's done. See my comment in the pull request.

**Comment by lvboque** [comment]
To simon for code review

**Comment by simjan** [comment]
Code ok for me... merge has been done.

**Comment by dergo** [comment]
I reopen the issue because of the following remaining issues:

1/ selectNode() call should fail if node cannot be selected (requested in points 1 and 7). (and test is supposed to check this, as documented, but it doesn't !)

2/ there are still some calls to System.out.println(), and please fix it also for other JavaGUI classes.

3/ In SelectionInTree, there is bad test doc: @SHOULD_FAIL should be @data SHOULD_FAIL.

4/ SelectionInTreeWithRegex is not complete: documentation is copy-pasted and regex feature is not completely tested, one just verify that "something" is selected, and in the test data one should use more regex characters at different places of the node path.

5/ In selectNodeRe(), replace the sentences "Be careful if you use some special regex characters.
Be careful to choose a node path separator which is not a special regex character." in the javadoc by "Be careful if you use some special regex characters in the node path separator". Note that you CAN use special regex characters in the node path separator, using e.g. selectNodeRe( 'TREE', 'Per.on\.Y[oa]ung', '\.')

**Comment by remybaranx** [comment]
I did some corrections directly in the master branch :

1.1/ selectNode() already failed if no node can be selected. If you had a problem, do you have an example where it doesn't fail ?

1.2/ I have splitted SelectionInTree test in 2 parts (normal cases and error cases). These tests will be completed.

2/ done for TreeNodeSelector.java. I will do this for others classes if I have time. Otherwise, create a new ticket for that.

3/ variable removed

4.1/ documentation updated.

4.2/ I have splitted SelectionInTreeWithRegex test in 2 parts (normal cases and error cases). These tests will be completed.

5.1/ sentence updated

5.2/ Yes you can use regex characters for a separator but, in your example, it will fail, so I don't understand the goal... For me, if you decide to use a regex character as separator, you have to be sure that this regex character is not used in your node path except to separate the node path elements.

Still to do:
- add test cases, mainly for selectNodeRe()
- remove System.out.println() calls.

PS: to answer to one of your remarks in a pull request about the selectNode() method rewriting : 
    + treeRootVisible() was used a lot of time without any reason.
    + there was no comment in the code.
    + I think the logic used to parse the tree was not clear at all and with some errors.
-> so I have decided to rewrite this function and I hope it's clearer now. I think sometimes it's better to rewrite than to patch. For the review, in this case, I think it's easier to read the function (do a code review ...) instead of comparing all the changes.

**Comment by dergo** [comment]
For 1.1/ and 1.2/, it does NOT work, selectNode doesn't fail but returns None if the node doesn't exist, and your new tests SelectionInTreeException/SelectionInTreeWithRegexException are not correct: if there is no exception, the tests pass. (you can verify this by replacing the 'time.sleep(1)' line at the end (which is btw useless) by 'testAPI.stopTest(Status.FAIL, "No exception")')

For 5.2/, it's because github removed the backslashes, here is the working code, with the backslashes:
```selectNodeRe( 'TREE', 'Per.on\.Y[oa]ung', '\.')```

Didn't had the time to review the other points.

**Comment by remybaranx** [comment]
I think it's ok now.
I have added a test case for selectNodeRe() with some different regex characters or expressions.

**Comment by simjan** [comment]
Code seems ok
all tests are green
 => validated
